codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -186,35 +184,31 @@ void onUnknownFrame0(ChannelHandlerContext ctx, byte frameType, int streamId, Ht
         @Override
         public int onDataRead(final ChannelHandlerContext ctx, int streamId, ByteBuf data,
                 int padding, boolean endOfStream) throws Http2Exception {
-            // Check if we received a data frame for a stream which is half-closed
             Http2Stream stream = connection.requireStream(streamId);
 
-            verifyGoAwayNotReceived();
+            if (stream.isResetSent() || streamCreatedAfterGoAwaySent(stream)) {
+                int bytesToReturn = data.readableBytes() + padding;
+                // Count the frame towards the connection flow control window and don't process it further.
+                flowController().receiveFlowControlledFrame(ctx, stream, data, padding, endOfStream);
```

Comment:
Move the `Http2LocalFlowController flowController = flowController();` up in this method and use it here.